File: src/Api/Helpers/JWT.cs

namespace Api.Helpers;
using System.Security.Claims;
using System.IdentityModel.Tokens.Jwt;
using System;
using Microsoft.IdentityModel.Tokens;
using System.Text;

public static class JWT {

    public static string CreateToken(string organization, string email, string secret) {
        Claim[] claims = [new Claim("organization", organization), new Claim("email", email)];
        var now = DateTime.UtcNow;
        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(secret));
        var creds = new SigningCredentials(key, SecurityAlgorithms.HmacSha256Signature);
        var jwtToken = new JwtSecurityToken(claims: claims,
                                            notBefore: now,
                                            expires: now.AddMinutes(60),
                                            signingCredentials: creds);
        return new JwtSecurityTokenHandler().WriteToken(jwtToken);
    }
}

File: src/Api/Helpers/JWT.fs
module JWT
open System.Security.Claims
open System.IdentityModel.Tokens.Jwt
open System
open Microsoft.IdentityModel.Tokens
open System.Text

let createToken (organization: string) (email: string) (secret: string) (durationMn: int) (settings: Api.AuthSettings)=
    let claims = [ 
        Claim("organization", organization)
        Claim("email", email)
    ]

    let startsOn = DateTime.UtcNow
    let expiresOn = startsOn.AddMinutes(durationMn)

    let key = SymmetricSecurityKey(Encoding.UTF8.GetBytes(secret))
    let creds = SigningCredentials(key, SecurityAlgorithms.HmacSha256Signature)
    let jwtToken = JwtSecurityToken(
        issuer = settings.Issuer,
        claims = claims,
        notBefore = startsOn,
        expires = expiresOn,
        signingCredentials = creds
    )
    JwtSecurityTokenHandler().WriteToken(jwtToken)

let getToken (auth: string) =
    JwtSecurityTokenHandler().ReadJwtToken(auth)

let getOrganization (token: JwtSecurityToken) =
    token.Claims |> Seq.tryFind (fun x -> x.Type = "organization")
